webhook/resources/engineimage/validator.go

```diff
@@ -0,0 +1,61 @@
+package engineimage
+
+import (
+	"github.com/pkg/errors"
+	admissionregv1 "k8s.io/api/admissionregistration/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/longhorn/longhorn-manager/datastore"
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+	"github.com/longhorn/longhorn-manager/types"
+	"github.com/longhorn/longhorn-manager/webhook/admission"
+	werror "github.com/longhorn/longhorn-manager/webhook/error"
+)
+
+type engineImageValidator struct {
+	admission.DefaultValidator
+	ds *datastore.DataStore
+}
+
+func NewValidator(ds *datastore.DataStore) admission.Validator {
+	return &engineImageValidator{ds: ds}
+}
+
+func (e *engineImageValidator) Resource() admission.Resource {
+	return admission.Resource{
+		Name:       "engineimages",
+		Scope:      admissionregv1.NamespacedScope,
+		APIGroup:   longhorn.SchemeGroupVersion.Group,
+		APIVersion: longhorn.SchemeGroupVersion.Version,
+		ObjectType: &longhorn.EngineImage{},
+		OperationTypes: []admissionregv1.OperationType{
+			admissionregv1.Create,
+			admissionregv1.Delete,
+		},
+	}
+}
+
+func (e *engineImageValidator) Create(request *admission.Request, newObj runtime.Object) error {
+	engineImage := newObj.(*longhorn.EngineImage)
+	if engineImage.Spec.Image == "" {
```

Comment:
Can we remove `// +optional` from `Image`? so we leave the validation to the OpenAPI scheme itself instead.

Reply:
`Image` is required field and non-empty sting.
Remove `// +optional` and Add `// +kubebuilder:validation:MinLength:=1`

Reply:
yeah, right now it's based on the error template of openAPI schema validation.